### ISSUE TITLE
Stuck at 0kb after deleting .mtd file

### ISSUE THREAD

**Issue by diptim01** [FaisalUmair/udemy-downloader-gui#330]
![Screen Shot 2019-11-14 at 2 59 32 AM](https://user-images.githubusercontent.com/19951351/68820397-6dc3f100-068b-11ea-9f52-c536a2d05a1c.png)

**Comment by FaisalUmair** [comment]
Just making sure you are following these steps.
Quit the program, delete the file, start Udeler. Click the download button again.   And also try disabling auto retry feature (if it is enabled)

**Comment by diptim01** [comment]
Done that. Still the same error on iOs and android

**Comment by FaisalUmair** [comment]
What do you mean by iOS and android?  The program has nothing to do with iOS or Android.

**Comment by diptim01** [comment]
> What do you mean by iOS and android? The program has nothing to do with iOS or Android.
> 
> Do you mean macOS ?

Oh sorry. I mean to say Mac and Windows

**Comment by FaisalUmair** [comment]
Have you tried to download any other course? And what version of Udeler are you using?

**Comment by diptim01** [comment]
V.1.8

It works very well before.  I'll reinstall the application and retry.

**Comment by FaisalUmair** [comment]
Do you only have one course in your account? Have you tried to download any other course. Just asking so that I know if it is a course specific issue or an issue with your computer or network.

**Comment by diptim01** [comment]
I'll retry and revert back to you. Thanks

**Comment by RamosCruzJeanCarlos** [comment]
I have the same problem days ago, I tried the steps they indicated, I reinstalled the udeler, I deleted the file that came out with Zero bytes, but it still doesn't download beyond that. The version I have is the latest v1.8.
I tried downloading other courses and you can download them, only the problem happens with a single course that was halfway down.
I tried deleting more files and what it does download is the .srt file, but Udeler no longer downloads any single videos of that course despite having downloaded before.

**Comment by Chris18coleman** [comment]
No solution yet about this??? I also have the same problem with multiple courses.

**Comment by DanMcLaughlin** [comment]
Yeah I'm seeing it also. I have previously downloaded 4 course on v1.8 no problem. Yesterday and today again I tried downloading a fifth course and it hangs early in the process, either at 1%, 5% or 7% from what I've seen so far. I tried restarting the app and deleting the initial files. 

I'm wondering if something changed on the udemy side.

**Comment by FaisalUmair** [comment]
@Chris18coleman @DanMcLaughlin  @RamosCruzJeanCarlos 

I understand this issue is quite common and whatever the reason is it leads to a failed download.

Now let me explain a bit what could be the reasons and what are some possible solutions. I want to make it very clear that not all of your course downloads are failing because of one particular problem. Here are some of the possible reasons and their solutions.

1) Due to expired download link which is stored in .mtd file. This one is very common.

**Solution**:  Delete the very last .mtd file or any other mtd file in the course folder which Udeler created for that course.


2) The course might have some content in it which Udeler fails to process. I mean there could be a file or some content for which the program doesn't know what to do and it fails to move ahead. 

**Solution**:  You need to send me your udemy credentials (over email, not here) so that I can try it and see what makes it fail and fix the issue and I am sure that will make lot of other users download their course as the reason could be similar.

3) The settings in program, could prevent it from downloading the lectures. For example if "Lecture Start/Lecture End" is enabled and you are not aware of it, it could prevent the course from starting download if the lecture start/end  is invalid or out of range for that course. And another issue related to settings could be the download path set earlier is no longer present or is unavailable or not writable. Ther could also be some other setting related issues, like the settings file (saved on your computer) have some bad values, is not readable and the settings you see in the program are showing you default values because the file settings didn't load.  (These are all my assumptions)

**Solution**: What you can do in this case is tweaking the settings and saving it again, forcing the file to get replaced. Changing download path or resetting it.

4) The request by the program are being cached and even if there is no mtd file present, the program is still being given the old data which contains old links. The reason I am not sure about this and still keeping it in the list is that this program is electron based which use chromium browser and I am using the browser as a client to make request to Udemy to get data. It is very unlikely that the data given to program is old considering the response headers returned by the API, but again it entirely depends on the browser (in our case chromium). And ironically one user recently contacted me regarding the same download issue, and he was able to fix it by deleting the app data of the program which would get rid of all the settings (which I set to saved in the default app data folder) configs and cache. So I really suggest you to try this if nothing else works.

**Solution**: Delete the contents of the folder the program uses to store settings, cache.

**MacOS**

If you're running macOS, your app's default user data directory is `~/Library/Application\ Support/<Udeler>`

**Windows**

If you're running Windows, your app's default user data directory is `%APPDATA%/<Udeler>`

**Linux**

If you're running Linux, your app's default user data directory is either `$XDG_CONFIG_HOME/<Udeler>` or `~/.config/<Udeler>`

**Comment by mauricioargj** [comment]
> @Chris18coleman @DanMcLaughlin @RamosCruzJeanCarlos
> 
> I understand this issue is quite common and whatever the reason is it leads to a failed download.
> 
> Now let me explain a bit what could be the reasons and what are some possible solutions. I want to make it very clear that not all of your course downloads are failing because of one particular problem. Here are some of the possible reasons and their solutions.
> 
> 1. Due to expired download link which is stored in .mtd file. This one is very common.
> 
> **Solution**: Delete the very last .mtd file or any other mtd file in the course folder which Udeler created for that course.
> 
> 1. The course might have some content in it which Udeler fails to process. I mean there could be a file or some content for which the program doesn't know what to do and it fails to move ahead.
> 
> **Solution**: You need to send me your Udemy credentials (over email, not here) so that I can try it and see what makes it fail and fix the issue and I am sure that will make lot of other users download their course as the reason could be similar.
> 
> 1. The settings in program, could prevent it from downloading the lectures. For example if "Lecture Start/Lecture End" is enabled and you are not aware of it, it could prevent the course from starting download if the lecture start/end  is invalid or out of range for that course. And another issue related to settings could be the download path set earlier is no longer present or is unavailable or not writable.  More on this here #331
>    There could also be some other setting related issues, like the settings file (saved on your computer) have some bad values, is not readable and the settings you see in the program are showing you default values because the file settings didn't load. (These are all my assumptions)
> 
> **Solution**: What you can do in this case is tweaking the settings and saving it again, forcing the file to get replaced. Changing download path or resetting it.
> 
> 1. The request by the program are being cached and even if there is no .mtd file present, the program is still being given the old data which contains old links. The reason I am not sure about this and still keeping it in the list is that this program is electron based which use chromium browser and I am using the browser as a client to make request to Udemy to get data. It is very unlikely that the data given to program is old considering the response headers returned by the API, but again it entirely depends on the browser (in our case chromium). And ironically one user recently contacted me regarding the same download issue, and he was able to fix it by deleting the app data of the program which would get rid of all the settings (which I set to saved in the default app data folder) configs and cache. So I really suggest you to try this if nothing else works.
> 
> **Solution**: Delete the contents of the folder the program uses to store settings, cache.
> 
> **MacOS**
> 
> If you're running macOS, your app's default user data directory is `~/Library/Application\ Support/<Udeler>`
> 
> **Windows**
> 
> If you're running Windows, your app's default user data directory is `%APPDATA%/<Udeler>`
> 
> **Linux**
> 
> If you're running Linux, your app's default user data directory is either `$XDG_CONFIG_HOME/<Udeler>` or `~/.config/<Udeler>`

I did it all, but nothing works, I still have the same problem.

**Comment by diptim01** [comment]
Maybe you should check the application log. Check each instance that
failed. That is the best way to troubleshoot it.
I believe there are many instances from your users.

**Comment by FaisalUmair** [comment]
@manuelbachl  As I said this could also be a course specific issue. In that case you would have to send me your Udemy credentials (over email) and course name. I will see if the download fails for me also, and in case it does I will fix it accordingly.

@diptim01  I don't have anything like that setup yet.

**Comment by Eli963741** [comment]
Hi @mauricioargj @FaisalUmair , 

no matter what I do, I'm not able to download course. I am downloading not popular, almost brand new course.  Just to mention it upfront front. 

I have installed, uninstalled, cleared cache almost milion times and nothing helps. I'm on Win10, Udeler 1.8. Changing IP adress via Nord VPN not helped at all, running Udeler as a Admin or adding Udeler to my Esset Firewall rules wouldn't work. 

![001](https://user-images.githubusercontent.com/56111013/68970699-27d06f80-07e8-11ea-80cf-f3e275732a29.PNG)
![002](https://user-images.githubusercontent.com/56111013/68970718-39197c00-07e8-11ea-9003-5df25fc3991b.PNG)

**Comment by Eli963741** [comment]
And this is 100% downloaded course.
![003](https://user-images.githubusercontent.com/56111013/68972995-674d8a80-07ed-11ea-8d03-1247aa38c7ec.PNG)

**Comment by Eli963741** [comment]
This is what is inside that 1k file
![005](https://user-images.githubusercontent.com/56111013/68978798-1d1fd580-07fc-11ea-9e9a-5749eebaae48.PNG)

**Comment by bomican** [comment]
Please fix ☺

**Comment by FaisalUmair** [comment]
@Eli963741 I would love to fix this issue and help you download the course. I need to debug the issue and find the root cause and for that I need your help.  First I would have to see if it is a course only issue and for that you have to provide me your Udemy credentials (you can send that over email).   If it was a  course only issue (problem with that particular issue) then I will it accordingly. You can also confirm that by yourself, by just checking if all your courses are failing or if you are able to download some of them.

Another way I can help you if you are not comfortable sharing the credentials is via remote desktop access.  This issue seems to be complicated, and I have no access to logs or any such info that can help. So I would need check myself and remote desktop access would be fine for me.


*If any one else here is having the same issue and willing to help in either of the above two ways to debug the issue, then let me know.

**Comment by Eli963741** [comment]
Hello @FaisalUmair thank you so much for helping me to solve my issue. I will send you and email with all details needed.

**Comment by FaisalUmair** [comment]
@Eli963741 Sure, I will continue the conversation there, and if we get it working, we can share the solution here.

**Comment by mauricioargj** [comment]
> @Eli963741 Sure, I will continue the conversation there, and if we get it working, we can share the solution here.


Any news @FaisalUmair? Thanks

**Comment by FaisalUmair** [comment]
@mauricioargj  I am still in touch with him. One thing I got to know that the course download starts for him but it gets stuck at a certain step.  My guess is that the particular course has many videos lectures and  since Udeler generates all the download before starting download (which is actually bad and I will change this in future), so sometimes because of slow internet connection, when it reaches to download a certain lecture, the download link stands expired and it fails. 


I am still debugging this issue, one thing is certain that not all of your download issues have same root cause. But I am sure making it generate a link one after another will make the program efficient anyway and might solve the issue for many.

**Comment by DanMcLaughlin** [comment]
@FaisalUmair The course causing trouble for me has much less content than other courses I've successfully downloaded. It also gets stuck at some step, but can start successfully. Thanks -

**Comment by FaisalUmair** [comment]
@DanMcLaughlin Thats why I said, not all issue are related. It could be that the course has some content which makes the program fail.  Only way to confirm that is to try Udeler on a different computer if it fails for that course, then you got the answer, and then you can share your credentials with me so that I can fix the issue after knowing where it fails and why.

However if it works on other computer, then it means that something is wrong with your instance, may be corrupt settings or that mtd issue I wrote about above, or network cache issue.   Anyway we have to first confirm it is not a course specific issue, for that you can also let me try download that course.

**Comment by mauricioargj** [comment]
> @mauricioargj Ainda estou em contato com ele. Uma coisa que eu pude saber é que o download do curso começa para ele, mas fica preso em uma determinada etapa. Meu palpite é que o curso em particular tem muitas palestras em vídeos e, como o Udeler gera todo o download antes de iniciar o download (o que é realmente ruim e vou mudar isso no futuro), às vezes por causa da conexão lenta à Internet, quando chega para baixar um certo palestra, o link para download permanece expirado e falha.
> 
> Ainda estou depurando esse problema. Uma coisa é certa: nem todos os problemas de download têm a mesma causa. Mas tenho certeza de que gerar um link um após o outro tornará o programa eficiente de qualquer maneira e poderá resolver o problema para muitos.



> @mauricioargj I am still in touch with him. One thing I got to know that the course download starts for him but it gets stuck at a certain step. My guess is that the particular course has many videos lectures and since Udeler generates all the download before starting download (which is actually bad and I will change this in future), so sometimes because of slow internet connection, when it reaches to download a certain lecture, the download link stands expired and it fails.
> 
> I am still debugging this issue, one thing is certain that not all of your download issues have same root cause. But I am sure making it generate a link one after another will make the program efficient anyway and might solve the issue for many.

In fact, the course I'm having trouble has many small parts of videos. It crashes early on, but the other courses I'm getting to download without any problem. Thank you so much for your effort and for your great work.

**Comment by mauricioargj** [comment]
> @mauricioargj I am still in touch with him. One thing I got to know that the course download starts for him but it gets stuck at a certain step. My guess is that the particular course has many videos lectures and since Udeler generates all the download before starting download (which is actually bad and I will change this in future), so sometimes because of slow internet connection, when it reaches to download a certain lecture, the download link stands expired and it fails.
> 
> I am still debugging this issue, one thing is certain that not all of your download issues have same root cause. But I am sure making it generate a link one after another will make the program efficient anyway and might solve the issue for many.

In fact, the course I'm having trouble has many small parts of videos. It crashes early on, but the other courses I'm getting to download without any problem. Thank you so much for your effort and for your great work.

**Comment by Chris18coleman** [comment]
Hey `FaisalUmair`, Any news about the interaction with `Eli963741`

**Comment by FaisalUmair** [comment]
@Chris18coleman I joined him on team viewer yesterday and today as well. I found some issues, which might be the reason for some of the download  fails, but for him it is a different issue that I also faced today while downloading his course. There a lot of things going on at the moment,  honestly there are multiple reasons why the download could be failing, so I will be fixing those which I found and will try to fix  the issue with his course also which might me the same issue some other users are facing / is the reason of their downloads failing.

**Comment by Chris18coleman** [comment]
> @Chris18coleman I joined him on team viewer yesterday and today as well. I found some issues, which might be the reason for some of the download fails, but for him it is a different issue that I also faced today while downloading his course. There a lot of things going on at the moment, honestly there are multiple reasons why the download could be failing, so I will be fixing those which I found and will try to fix the issue with his course also which might me the same issue some other users are facing / is the reason of their downloads failing.

Awesome! when do you think this tests and fixes gonna be ready? I mean when you plan to post these solutions that you found due to this interactions?

Thanks in advanced!

**Comment by FaisalUmair** [comment]
> > @Chris18coleman I joined him on team viewer yesterday and today as well. I found some issues, which might be the reason for some of the download fails, but for him it is a different issue that I also faced today while downloading his course. There a lot of things going on at the moment, honestly there are multiple reasons why the download could be failing, so I will be fixing those which I found and will try to fix the issue with his course also which might me the same issue some other users are facing / is the reason of their downloads failing.
> 
> Awesome! when do you think this tests and fixes gonna be ready? I mean when you plan to post these solutions that you found due to this interactions?
> 
> Thanks in advanced!

I won't be getting enough time to work on this until weekend, so that will slow down the progress. But I am keen on fixing this issue and all download related issues.  It will require some time from me to work on it and fix it completely, it seems to be a bit complex right now, so might require me to redo some part of the program. I can't promise anything as of now, but will try to address most of these download issues and have an update ready by the end of next week.

**Comment by DanMcLaughlin** [comment]
Interesting change - on the course that wasn't downloading, now it started and got 10/43 Into the download before hanging. Before it would hang on the first one.

**Comment by thornzz** [comment]
I have got same problem. Whats your e-mail address i can mail you my udemy credentials.

**Comment by RamosCruzJeanCarlos** [comment]
> > > @Chris18coleman I joined him on team viewer yesterday and today as well. I found some issues, which might be the reason for some of the download fails, but for him it is a different issue that I also faced today while downloading his course. There a lot of things going on at the moment, honestly there are multiple reasons why the download could be failing, so I will be fixing those which I found and will try to fix the issue with his course also which might me the same issue some other users are facing / is the reason of their downloads failing.
> > 
> > 
> > Awesome! when do you think this tests and fixes gonna be ready? I mean when you plan to post these solutions that you found due to this interactions?
> > Thanks in advanced!
> 
> I won't be getting enough time to work on this until weekend, so that will slow down the progress. But I am keen on fixing this issue and all download related issues. It will require some time from me to work on it and fix it completely, it seems to be a bit complex right now, so might require me to redo some part of the program. I can't promise anything as of now, but will try to address most of these download issues and have an update ready by the end of next week.

Thank you very much for the help, I found a temporary solution, which consists of connecting to another internet network, I connected to another Wi-Fi network and I could download with the course that had stayed at 0kb / s try to try suddenly it can work for you
![Captura de Pantalla 2019-11-24 a la(s) 00 13 52](https://user-images.githubusercontent.com/19050069/69490058-e4ad7680-0e4f-11ea-97f4-d51d7d23fc9f.png)

**Comment by SametTurgut** [comment]
I have the same issue.
